content/posts/2022-02-02-decreased-backend-tests-step/index.md

```diff
@@ -0,0 +1,174 @@
+
+After mocking properly them, we could have an improvement of 35% by mocking 1,84% tests properly.
+
+In order to prevent it to happen again, we left [pytest-socket](https://github.com/miketheman/pytest-socket) properly configured in our codebase.
+
+4. Upgrade dependencies (including pytest)
+
+When we were adding [pytest-socket](https://github.com/miketheman/pytest-socket), we noticed that most of the project dependencies were outdated.
+Hypothesis: Using the latest project dependencies version could have performance improvements?
+
+It was a huge effort to update them all and make sure that everything was working as expected. For example, by upgrading pytest from version 2 to version 6, we had to change some decorator orders, like `@freeze_time` needed to be placed after the `@pytest.fixture` decorator.
+
+By the end of this effort, we could not see any performance improvement. However, it was really valuable to have all dependencies updated, since the latest version could include performance and security fixes.
+
+5. Running tests in parallel
+
+After all those tentatives, we tried to back to the initial investigation results. By looking at CircleCI documentation, we found the feature of [running tests in parallel](https://circleci.com/docs/2.0/parallelism-faster-jobs/).
+
+CircleCI provides the option to split tests by timing, which was exactly what we were looking for. By doing so, we had our big moment: the test step decreased from about 22m to 6m.
+
+However, Coveralls started failing, since each job had a portion of coverage data. We had to combine multiple coverage files into a single one:
+
+```yaml
+    # ...
+    - run:
+    name: Run Python tests
+    command: |
+        TEST_FILES=$(circleci tests glob "head/**/test_*.py" | circleci tests split --split-by=timings)
+        pytest --no-migrations --junitxml=build/report_${CIRCLE_NODE_INDEX}.xml --cov-append --cov=head --cov-config .coveragerc --cov-report term --cov-report xml --capture=no -n 15 $TEST_FILES
+    - run: |
+        mv build/coverage.xml build/coverage_${CIRCLE_NODE_INDEX}.xml
+        mv .coverage build/.coverage_${CIRCLE_NODE_INDEX}
+    - store_test_results:
+        path: build
+    #...
+    - run:
+    name: Publish coverage
+    command: |
+        coverage combine build/.coverage_*
+        coveralls
+    #...
+```
+
+### Bonus action:
+
+CircleCI also provides a great feature to [cache the docker layer](https://circleci.com/docs/2.0/docker-layer-caching/). It may give you some more time improvements.
+We did not take a look at it because we were already using it.
+
+
+## Lessons learned
+
+- Tech debts should not be ignored and fixes are WORTH to invest.
+
+    In this particular large codebase, we had several tech debts raised by this Working Group. Some of them we could fix, but it was something we as engineers should keep our eyes open to avoid it existing for that long. As time goes by, it may become costly to pay.
```

Comment:
```suggestion
In this particular large codebase, we had several tech debts that were found by this Working Group. Some of them we were able to fix, but this is something we as engineers should keep our eyes on, to avoid it existing for that long. As time goes by, it may become costly to pay those debts.
```